fix panic error when docker stats a stopped container

hqhq opened the pull request:
Signed-off-by: Qiang Huang h.huangqiang@huawei.com

cpuguy83 commented:
Oh shoot, I thought this was fixed in #11208.
LGTM

estesp commented:
Seems this would be easy to add a test, and catches a very ugly error (panic).

hqhq commented:
Yes, something wrong happened on docker stats stopped container before, now it happen again, it's necessary we have such a test case.
It's a bit difficult to test stats command, so I add a test case for stats REST api.

estesp commented on integration-cli/docker_api_containers_test.go:
```diff
@@ -320,6 +320,49 @@ func TestGetContainerStats(t *testing.T) {
+	time.Sleep(4 * time.Second)
+	if _, err := runCommand(exec.Command(dockerBinary, "rm", "-f", name)); err != nil {
+		t.Fatal(err)
+	}
```
This section is not needed as for your test there will not be anything returned from `/stats` endpoint; and your test is to validate that it doesn't panic at this point (without your fix)!  This adds time to the tests that isn't really valuable, so would be better to remove this whole block and add rely on the `deleteAllContainers()` you already have at the beginning of your test function.

estesp commented:
Thanks for adding the test--I made one comment on a modification to the test.  After that can you squash this into one commit, and we'll be ready to merge.

hqhq committed:
fix panic error when docker stats a stopped container

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@estesp Thanks for your advice, updated test case.
When I modified the test case, I found sockRequest("GET", "containers/"+name+"/stats", nil) would never return, I think it was because ioutil.ReadAll() will never return because we won't get error or EOF from daemon when we GET stats. So for the former test TestGetContainerStats, we'll never get any really body but always wait for the rm command to stop and remove the container.

ping @crosbymichael @LK4D4 see if we have some better choice for stats test cases.

crosbymichael commented:
LGTM
